src/transformers/models/gemma/modeling_gemma.py

```diff
@@ -864,13 +863,24 @@ def forward(
                 past_key_values = DynamicCache.from_legacy_cache(past_key_values)
             past_seen_tokens = past_key_values.get_seq_length()
 
-        if cache_position is None:
-            cache_position = torch.arange(
+        if position_ids is None:
+            if isinstance(past_key_values, StaticCache):
+                raise ValueError("position_ids is a required argument when using StaticCache.")
+            position_ids = torch.arange(
                 past_seen_tokens, past_seen_tokens + inputs_embeds.shape[1], device=inputs_embeds.device
-            )
+            ).unsqueeze(0)
 
-        if position_ids is None:
-            position_ids = cache_position.unsqueeze(0)
+        # One of the rows in `position_ids` contains the highest sequence of cache indexes, excluding left-padding
+        # applied on all batch members. Left-padding on all batch members can be detected from the `attention_mask`.
+        cache_position = torch.max(position_ids, dim=0).values
+        if attention_mask is None:
+            padded_offset = 0
+        else:
+            padded_offset = (1 - torch.sum(attention_mask.to(torch.int64), dim=0).clamp(max=1)).cumsum(-1)
+            padded_offset = torch.cat(
+                (torch.zeros((1,), dtype=torch.int64, device=padded_offset.device), padded_offset)
+            )[-cache_position.shape[0] - 1 : -1]
+        cache_position = cache_position + padded_offset
```

Comment:
you are using sum + clamp + cumsum + cat , costly for : long sequence length, 4D attention mask provided to the model.

Reply:
Not quite, see performance numbers above :)